module/spock/spock_verifier.go

```diff
@@ -0,0 +1,191 @@
+		approvals:     make(map[flow.Identifier]map[flow.Identifier]*flow.ExecutionReceipt),
+	}
+}
+
+// AddReceipt adds a receipt into map if the SPoCK proofs do not match any other receipt
+func (v *Verifier) AddReceipt(receipt *flow.ExecutionReceipt) error {
+	resultID := receipt.ExecutionResult.ID()
+
+	// if receipts result id does not exist in map, create an array with the receipt
+	_, ok := v.receipts[resultID]
+	if !ok {
+		receipts := make([]*flow.ExecutionReceipt, 0)
+		receipts = append(receipts, receipt)
+		v.receipts[resultID] = receipts
+		return nil
+	}
+
+	// check if candidate receipt SPoCK proofs match proofs in map
+	matched, err := v.matchReceipt(receipt)
+	if err != nil {
+		return fmt.Errorf("could not match receipt: %w", err)
+	}
+
+	// if not matched add to receipt array
+	if !matched {
+		v.receipts[resultID] = append(v.receipts[resultID], receipt)
+	}
+
+	// if matched we do nothing (strong-transitive property of the SPoCK scheme)
+
+	return nil
+}
+
+// VerifyApproval verifies an approval with all the distinct receipts for the approvals
+// result id and returns true if spocks match else false
+func (v *Verifier) VerifyApproval(approval *flow.ResultApproval) (*flow.ExecutionReceipt, error) {
+
+	// check if verification already done and that a match is there
+	// note: we do not add to cache if the approval was not verified
+	receipt, ok := v.approvals[approval.Body.ExecutionResultID][approval.ID()]
+	if ok {
+		return receipt, nil
+	}
+
+	// find identities
+	approver, err := v.protocolState.AtBlockID(approval.Body.BlockID).Identity(approval.Body.ApproverID)
+	if err != nil {
+		return nil, fmt.Errorf("could not find approver identity")
+	}
+
+	receipts := v.receipts[approval.Body.ExecutionResultID]
+	for _, receipt := range receipts {
+		executor, err := v.protocolState.AtBlockID(receipt.ExecutionResult.BlockID).Identity(receipt.ExecutorID)
+		if err != nil {
+			return nil, fmt.Errorf("could not find executor identity")
+		}
+
+		// verify spock
+		verified, err := crypto.SPOCKVerify(approver.StakingPubKey, approval.Body.Spock, executor.StakingPubKey, receipt.Spocks[approval.Body.ChunkIndex])
+		if err != nil {
+			return nil, fmt.Errorf("could not verify spocks: %w", err)
+		}
+		if verified {
+			// map doesnt exist create it
+			if _, ok = v.approvals[approval.Body.ExecutionResultID]; !ok {
+				v.approvals[approval.Body.ExecutionResultID] = make(map[flow.Identifier]*flow.ExecutionReceipt)
+			}
+
+			// cache result for future verify calls for this approval
+			v.approvals[approval.Body.ExecutionResultID][approval.ID()] = receipt
+
+			return receipt, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// ClearReceipts clears all receipts for a specific resultID
+func (v *Verifier) ClearReceipts(resultID flow.Identifier) bool {
+	v.Lock()
+	defer v.Unlock()
+
+	// check if entry exists
+	_, ok := v.receipts[resultID]
+	if !ok {
+		return false
+	}
+	_, ok = v.approvals[resultID]
+	if !ok {
+		return false
+	}
```

Comment:
Not sure if it is a good idea to remove approval for a result. I think we cache results in `approvals` for future calls to  `VerifyApproval` and to avoid re-doing the same job again, right? then shouldn't we let `resultID` reside in the `approvals`? Otherwise, we are making a stateless verifier, which I think we are not exactly aiming for. 

```
			// cache result for future verify calls for this approval
			v.approvals[approval.Body.ExecutionResultID][approval.ID()] = receipt

```